src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,259 @@
+"""Low level control logic for actuating the rudder and the sail."""
+
+from math import atan2, copysign, cos, sin
+from typing import Tuple
+
+import numpy as np
+
+from boat_simulator.common.constants import RUDDER_MAX_ANGLE_RANGE, SAIL_MAX_ANGLE_RANGE
+from boat_simulator.common.types import Scalar
+from boat_simulator.common.utils import bound_to_180
+
+
+class ActuatorController:
+    """Abstract class for rudder and sail actuation mechanisms.
+
+    Attributes:
+        `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+        `time_step` (Scalar): Time taken per iteration given in seconds.
+        `control_speed` (Scalar): Speed of control angle change in degrees / second.
+        'running_error' (Scalar): Error between current and target control angle.
+    """
+
+    def __init__(
+        self,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        control_speed: Scalar,
+        max_angle_range: Tuple[Scalar, Scalar],
+    ):
+        """Initializes the class attributes. Note that this class cannot be directly instantiated.
+
+        Args:
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in seconds.
+            `control_speed` (Scalar): Speed of control angle change in degrees / second.
+            'running_error' (Scalar): Error between current and target control angle. Default to 0
+        """
+        self.current_control_ang = current_control_ang
+        self.time_step = time_step
+        self.control_speed = control_speed
+        self.max_angle_range = max_angle_range
+        self.running_error = 0.0
+
+    def update_state(self) -> bool:
+        """Updates the controller position iteratively based on control_speed * time_step, until
+        the control target angle is reached.
+
+        Args:
+            `max_angle_range` (tuple): Max angle range individual to each different controller
+
+
+        Returns:
+            False if target control angle was not reached, else
+            True if target control angle has been reached (running error = 0)
+        """
+        error_tracker = False
+        if np.isclose(self.running_error, 0, 0.01):
+            return True
+        else:
+            change = copysign(self.control_speed * self.time_step, self.running_error)
```

Comment:
A few suggestions:

1. Rename `error_tracker` to something more descriptive to its purpose, like `is_target_reached`
2. Have the `if` statement be the first thing you do, however, don't include an `else` branch and just compute the change outside the `if` statement. If the `if` statement evaluates to true, you just return anyways, so there is no need for an else branch. Most other languages would complain that you defined `change` in the else branch but use it outside because it could "technically be not defined at all".